src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -65,6 +65,30 @@ public static object GetUninitializedObject(Type type)
             return GetUninitializedObjectInternal(type);
         }
 
+        public static void ExecuteCodeWithGuaranteedCleanup(TryCode code, CleanupCode backoutCode, object userData)
+        {
+            if (code == null)
+                throw new ArgumentNullException(nameof(code));
+            if (backoutCode == null)
+                throw new ArgumentNullException(nameof(backoutCode));
+
+            bool exceptionThrown = false;
+
+            try
+            {
+                code(userData);
+            }
+            catch
+            {
+                exceptionThrown = true;
+                throw;
+            }
+            finally
+            {
+                backoutCode(userData, exceptionThrown);
+            }
```

Comment:
NIT: perhaps can initialize exceptionThrown to true and then set to false right before leaving try. No need for catch/rethrow then. Not sure if that is any better for execution, but it wold be shorter somewhat.